File: py/aiomavsdk/examples/camera-client.py
import asyncio
from aiomavsdk import Mavsdk, Configuration, ComponentType
from aiomavsdk.plugins.camera import CameraAsync, Mode


async def print_storage_updates(camera_plugin: CameraAsync):
    async for update in camera_plugin.subscribe_storage():
        print("Camera storage update:")
        print(f"  Component ID: {update.storage.component_id}")
        print(f"  Storage ID: {update.storage.storage_id}")
        print(f"  Total storage: {update.storage.total_storage_mib:.2f} MiB")
        print(f"  Used storage: {update.storage.used_storage_mib:.2f} MiB")
        print(f"  Available storage: {update.storage.available_storage_mib:.2f} MiB")
        print(f"  Storage status: {update.storage.storage_status}")
        print(f"  Storage type: {update.storage.storage_type}")


async def do_camera_operation(component_id: int, camera_plugin: CameraAsync):
    await camera_plugin.set_mode(component_id, Mode.PHOTO)
    print("Set camera to photo mode")

    await camera_plugin.take_photo(component_id)
    print("Took photo")

    await camera_plugin.set_mode(component_id, Mode.VIDEO)
    print("Set camera to video mode")

    await camera_plugin.start_video(component_id)
    print("Started video")

    await camera_plugin.stop_video(component_id)
    print("Stopped video")

    await camera_plugin.start_video_streaming(component_id, 0)
    print("Started video streaming")

    await camera_plugin.stop_video_streaming(component_id, 0)
    print("Stopped video streaming")

    await camera_plugin.format_storage(component_id, 0)
    print("Formatted storage")

    await camera_plugin.reset_settings(component_id)
    print("Reset camera settings")


async def main():
    configuration = Configuration.create_with_component_type(ComponentType.GROUND_STATION)
    mavsdk = Mavsdk(configuration)
    await mavsdk.add_any_connection("udpout://127.0.0.1:14030")
    print("Connected!")

    print("Waiting for camera system...")
    camera_server = None
    async for _ in mavsdk.on_new_system():
        systems = await mavsdk.get_systems()
        if systems:
            camera_server = systems[0]
            break

    camera_plugin = CameraAsync(camera_server)

    print("Waiting for cameras...")
    while True:
        camera_list = await camera_plugin.camera_list()
        if camera_list and len(camera_list.cameras) > 0:
            break
        await asyncio.sleep(1)

    print("Cameras found:")
    for camera in camera_list.cameras:
        print(f"Camera component ID: {camera.component_id}, model: {camera.model_name}, vendor: {camera.vendor_name}")

    if len(camera_list.cameras) > 1:
        print("More than one camera found, using first one discovered.")

    component_id = camera_list.cameras[0].component_id

    storage_task = asyncio.create_task(print_storage_updates(camera_plugin))

    await do_camera_operation(component_id, camera_plugin)

    print("Camera operations completed. Keeping running to receive updates...")
    try:
        await asyncio.sleep(float('inf'))
    except asyncio.CancelledError:
        pass
    finally:
        storage_task.cancel()
        try:
            await storage_task
        except asyncio.CancelledError:
            pass


if __name__ == "__main__":
    asyncio.run(main())

File: cpp/examples/camera_server/camera_client.cpp
#include <chrono>
#include <future>
#include <format>
#include <iostream>
#include <thread>

#include <mavsdk/mavsdk.h>
#include <mavsdk/plugins/camera/camera.h>

static void do_camera_operation(int32_t component_id, mavsdk::Camera& camera);

int main(int argc, const char* argv[])
{
    // we run client plugins to act as the GCS
    // to communicate with the camera server plugins.
    mavsdk::Mavsdk mavsdk{mavsdk::Mavsdk::Configuration{mavsdk::ComponentType::GroundStation}};

    auto result = mavsdk.add_any_connection("udpout://127.0.0.1:14030");
    if (result == mavsdk::ConnectionResult::Success) {
        std::cout << "Connected!" << std::endl;
    }

    auto prom = std::promise<std::shared_ptr<mavsdk::System>>{};
    auto fut = prom.get_future();
    mavsdk::Mavsdk::NewSystemHandle handle =
        mavsdk.subscribe_on_new_system([&mavsdk, &prom, &handle]() {
            auto system = mavsdk.systems().back();

            if (system->has_camera()) {
                std::cout << "Discovered camera from Client" << std::endl;

                // Unsubscribe again as we only want to find one system.
                mavsdk.unsubscribe_on_new_system(handle);
                prom.set_value(system);
            } else {
                std::cout << "No MAVSDK found" << std::endl;
            }
        });

    if (fut.wait_for(std::chrono::seconds(10)) == std::future_status::timeout) {
        std::cout << "No camera found, exiting" << std::endl;
        return -1;
    }

    auto system = fut.get();

    auto camera = mavsdk::Camera{system};

    camera.subscribe_camera_list([](mavsdk::Camera::CameraList list) {
        std::cout << "Cameras found:" << std::endl;
        for (auto& camera : list.cameras) {
            std::cout << std::format("{}\n", camera);
        }
    });

    // Wait for camera to be discovered.
    std::this_thread::sleep_for(std::chrono::milliseconds(500));

    // We expect to find one camera.
    if (camera.camera_list().cameras.size() == 0) {
        std::cerr << "No camera found, exiting.\n";
        return 1;
    }

    if (camera.camera_list().cameras.size() > 1) {
        std::cout << "More than one camera found, using first one discovered.\n";
    }

    const auto component_id = camera.camera_list().cameras[0].component_id;

    camera.subscribe_storage([](mavsdk::Camera::StorageUpdate update) {
        std::cout << "Camera storage update:" << std::endl;
        std::cout << std::format("{}\n", update);
    });

    do_camera_operation(component_id, camera);

    // for test subscribe camera status, so don't end the process
    while (true) {
        std::this_thread::sleep_for(std::chrono::seconds(1));
    }
    return 0;
}

void do_camera_operation(int32_t component_id, mavsdk::Camera& camera)
{
    // switch to photo mode to take photo
    auto operation_result = camera.set_mode(component_id, mavsdk::Camera::Mode::Photo);
    std::cout << std::format("Set camera to photo mode result : {}\n", operation_result);

    operation_result = camera.take_photo(component_id);
    std::cout << std::format("Take photo result : {}\n", operation_result);

    // switch to video mode to recording video
    operation_result = camera.set_mode(component_id, mavsdk::Camera::Mode::Video);
    std::cout << std::format("Set camera to video mode result : {}\n", operation_result);

    operation_result = camera.start_video(component_id);
    std::cout << std::format("Start video result : {}\n", operation_result);

    operation_result = camera.stop_video(component_id);
    std::cout << std::format("Stop video result : {}\n", operation_result);

    // the camera can have multi video stream so may need the stream id
    operation_result = camera.start_video_streaming(component_id, 0);
    std::cout << std::format("start video streaming result : {}\n", operation_result);

    operation_result = camera.stop_video_streaming(component_id, 0);
    std::cout << std::format("stop video streaming result : {}\n", operation_result);

    // format the storage with special storage id test
    operation_result = camera.format_storage(component_id, 0);
    std::cout << std::format("format storage result : {}\n", operation_result);

    operation_result = camera.reset_settings(component_id);
    std::cout << std::format("Reset camera settings result : {}\n", operation_result);
}
